rtrlib/pfx_table.py

```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf8 -*-
+"""
+rtrlib.rtr_manager
+------------------
+
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+
+from .exceptions import PFXException
+from .rtr_manager import ValidationResult
+from .util import ip_str_to_addr
+
+from _rtrlib import ffi, lib
+
+
+class PfxTable(object):
+    """
+    Wrapper class around pfx_table.
+
+    """
+
+    def __init__(self):
+        # allocate pfx_table
+        self.pfx_table = ffi.new('struct pfx_table *')
+        # initialize it
+        lib.pfx_table_init(self.pfx_table,
+                           ffi.NULL)
+        self.closed = False
+
+    def add_record(self, asn, ip, min_length, max_length):
+        """
+        Add a BGP prefix to the table.
+
+        :param asn: autonomous system number
+        :type asn: int
+
+        :param ip: ip address
+        :type ip: str
+
+        :param min_length: minimum length of the subnet mask
+        :type min_length: int
+
+        :param max_length: minimum length of the subnet mask
+        :type max_length: int
+        """
+
+        record = ffi.new('struct pfx_record *')
+        prefix = ffi.new('struct lrtr_ip_addr *')
+        lib.lrtr_ip_str_to_addr(ip.encode('ascii'), prefix)
+
+        record.asn = asn
+        record.min_len = min_length
+        record.max_len = max_length
+        record.socket = ffi.NULL
+        record.prefix = prefix[0]
+
+        lib.pfx_table_add(self.pfx_table, record)
+
+    def remove_record(self, asn, ip, min_length, max_length):
+        """
+        Add a BGP prefix to the table.
+
+        :param asn: autonomous system number
+        :type asn: int
+
+        :param ip: ip address
+        :type ip: str
+
+        :param min_length: minimum length of the subnet mask
+        :type min_length: int
+
+        :param max_length: minimum length of the subnet mask
+        :type max_length: int
+        """
+
+        record = ffi.new('struct pfx_record *')
+        prefix = ffi.new('struct lrtr_ip_addr *')
+        lib.lrtr_ip_str_to_addr(ip.encode('ascii'), prefix)
```

Comment:
You could do `lib.lrtr_ip_str_to_addr(ip.encode('ascii'), ffi.new("struct lrtr_ip_addr *", record.prefix))` instead of doing an additional alloc and copy.

Reply:
And maybe create a function like `create_pfx_record` to avoid duplication?

Reply:
Hm, for some reason the record would not get removed or added (lines 50-51 are similar) if I make your suggested change.

Reply:
However, while testing I spotted a weak point at class ValidationResult (line 471). If an object is instantiated without a reason, then this line will raise a TypeError.

`TypeError: expected a string or cdata object, got 'NoneType'`

Which is immediately followed by an AttributeError:

`AttributeError: 'ValidationResult' object has no attribute '_reason_records'`

Reply:
I added a small guard condition.

Reply:
The duplication issue is now resolved.

Reply:
Sorry, should have tested it before suggesting …
The correct call is `ffi.addressof(record, 'prefix')`

Reply:
Yes, after a short try&error I got this working.